src/main/java/com/iexec/blockchain/tool/QueueService.java

```diff
@@ -1,21 +1,122 @@
 package com.iexec.blockchain.tool;
 
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.scheduling.annotation.Scheduled;
 import org.springframework.stereotype.Service;
 
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.PriorityBlockingQueue;
 
+/**
+ * Execute {@link Runnable}s as they arrive.
+ * Some {@link Runnable}s can have higher priority than others, so they can be treated before the others.
+ */
+@Slf4j
 @Service
 public class QueueService {
-
     private final ExecutorService executorService;
+    private final PriorityBlockingQueue<WaitingTask> queue = new PriorityBlockingQueue<>();
+
+    private CompletableFuture<Void> taskExecutor;
 
     public QueueService() {
         executorService = Executors.newFixedThreadPool(1);
     }
 
-    public void runAsync(Runnable runnable) {
-        CompletableFuture.runAsync(runnable, executorService);
+    /**
+     * Scheduled method execution.
+     * If the {@link QueueService#taskExecutor} isn't running, start a new thread.
+     * Otherwise, do nothing.
+     */
+    @Scheduled(fixedRate = 30000)
+    private void startAsyncTasksExecution() {
+        if (taskExecutor != null && !taskExecutor.isDone()) {
+            return;
+        }
+        
+        taskExecutor = CompletableFuture.runAsync(this::executeTasks);
+    }
+
+    /**
+     * Execute {@link Runnable}s as they come.
+     * At each occurrence, execute the first task in the queue or wait for a new one to appear.
+     * The first task is the task with the highest priority that is in the queue for the longer time.
+     */
+    void executeTasks() {
+        while (Thread.currentThread().isAlive()) {
+            executeFirstTask();
+        }
+    }
+
+    /**
+     * Execute the first task in the queue or wait for a new one to appear.
+     * The first task is the task with the highest priority that is in the queue for the longer time.
+     */
+    void executeFirstTask() {
+        try {
+            // Wait until a new task is available.
+            Runnable runnable = queue.take().getRunnable();
+            CompletableFuture.runAsync(runnable, executorService);
```

Comment:
If I'm not wrong it looks you are fully emptying the PriorityBlockingQueue queue (made of prioritized runnable) and filling the ExecutorService queue..
We should wait current runnable completion before taking new ones.